### DISCUSSION TITLE
sipeed longan nano

### DISCUSSION THREAD

**Discussion by vorosj** [hathach/tinyusb#1787] [Q&A]
Make is not working for the longan nano. (working for other uC-s like stm32f103.
win11/wsl 
What to do?

mmormota@DESKTOP-S2K747J:/mnt/c/tinyusb/examples/device/board_test$ make BOARD=sipeed_longan_nano get-deps
git -C /mnt/c/tinyusb submodule update --init hw/mcu/gd/nuclei-sdk
mmormota@DESKTOP-S2K747J:/mnt/c/tinyusb/examples/device/board_test$ make BOARD=sipeed_longan_nano all
AS startup_gd32vf103_asm.o
/bin/sh: 1: riscv-none-embed-gcc: not found
make: *** [../../rules.mk:143: _build/sipeed_longan_nano/obj/hw/mcu/gd/nuclei-sdk/SoC/gd32vf103/Common/Source/GCC/startup_gd32vf103_asm.o] Error 127
mmormota@DESKTOP-S2K747J:/mnt/c/tinyusb/examples/device/board_test$

mmormota@DESKTOP-S2K747J:/mnt/c/tinyusb/examples/device/board_test$ make BOARD=stm32f103_bluepill all
AS startup_stm32f103xb_asm.o
CC dcd_stm32_fsdev.o
...
LINK _build/stm32f103_bluepill/board_test.elf
CREATE _build/stm32f103_bluepill/board_test.bin
CREATE _build/stm32f103_bluepill/board_test.hex

   text    data     bss     dec     hex filename
   1844       8    1544    3396     d44 _build/stm32f103_bluepill/board_test.elf

**Comment by cr1901** [comment]
GCC binaries targeting RISCV do not have a standardized name, even for bare metal. Edit the [linked file](https://github.com/hathach/tinyusb/blob/master/hw/bsp/gd32vf103/family.mk#L1-L8) in your copy of the repo choose a different toolchain name from `riscv-none-embed-gcc`, depending on what RISCV toolchain name you actually have installed.

Alternatively, I think `make CROSS_COMPILE=riscv32-unknown-elf- BOARD=sipeed_longan_nano all` will also work (untested).

**Reply by vorosj** [reply]
Unfortunately the alternative command is not working.

There is a nuclei sdk folder in my tinyusb folder:
C:\tinyusb\hw\mcu\gd\nuclei-sdk
There is no gcc, gcc/bin etc folder, so maybe the toolchain is missing. I am not familiar with Linux, what is the proper wy to install it, if it's really missing?

I assumed the command 
$ make BOARD=sipeed_longan_nano get-deps
shuld install the toolchain, isn't it?

Sorry for the lamer questions...

**Reply by cr1901** [reply]
I don't believe a toolchain is supposed to be installed as part of `get-deps`- only submodules. It seems like a happy coincidence that you have an ARM toolchain already installed, but not RISCV.

I would suggest [installing](https://xpack.github.io/riscv-none-elf-gcc/install/) the `riscv-none-embed-gcc` tools. Alternatively, if you don't want to opt-into yet another package manager, I would [download](https://github.com/sifive/freedom-tools/releases/tag/v2020.12.0) SiFive's version of the prebuilt tools and untar (`tar -xvf`) them into `~/.local` or similar. Make sure `~/.local/bin` or whatever is on your path and you update `CROSS_COMPILE=riscv64-unknown-elf-` on your command-line or in the file I linked in the previous answer.

To @hathach: Is `CROSS_COMPILE` documented as user-changeable? The toolchain naming for RISCV `gcc` variants is inconsistent/bad. It just is. I don't actually have any of the toolchains [listed](https://github.com/hathach/tinyusb/blob/master/hw/bsp/gd32vf103/family.mk#L1-L8) in that file either; mine is `riscv64-unknown-elf-gcc`. Despite the name, `riscv64` is perfectly acceptable for bare-metal 32-bit RISCV dev with the right `march` [option](https://www.sifive.com/blog/all-aboard-part-1-compiler-args).

**Reply by vorosj** [reply]
Thank you very much.
I downloaded the prebuilt tool and untar them to C:\tinyusb\hw\mcu\gd\nuclei-sdk\gcc
checked, that the **riscv-nuclei-elf-gcc** file is in the gcc/bin folder.
I added a path: C:\tinyusb\hw\mcu\gd\nuclei-sdk\gcc\bin\
(and you are right, there was this path: c:\0work\gcc-arm-11.2-2022.02-mingw-w64-i686-arm-none-eabi\bin\)

Unfortunately I still get error:
mmormota@DESKTOP-S2K747J:/mnt/c/tinyusb/examples/device/uac2_headset$ make CROSS_COMPILE=riscv-nuclei-elf- BOARD=sipeed_longan_nano all
AS startup_gd32vf103_asm.o
make: riscv-nuclei-elf-gcc: Command not found
make: *** [../../rules.mk:143: _build/sipeed_longan_nano/obj/hw/mcu/gd/nuclei-sdk/SoC/gd32vf103/Common/Source/GCC/startup_gd32vf103_asm.o] Error 127
mmormota@DESKTOP-S2K747J:/mnt/c/tinyusb/examples/device/uac2_headset$

**Reply by cr1901** [reply]
You need to make sure `riscv-nuclei-elf-gcc` is on your system path (`export PATH=/path/to/nuclei/bin:$PATH` on Linux, `set PATH=/path/to/nuclei/bin;%PATH%` on Windows or similar). If you can't run `riscv-nuclei-elf-gcc` from your command-prompt like that, then the `make` command will fail.

**Reply by vorosj** [reply]
Thank you, that was fast... :-)
I am really lamer in Linux, no luck. I did this in the WSL Linux window:

mmormota@DESKTOP-S2K747J:/mnt/c/tinyusb/examples/device/uac2_headset$ export PATH=C:\tinyusb\hw\mcu\gd\nuclei-sdk\gcc\bin\:$PATH
mmormota@DESKTOP-S2K747J:/mnt/c/tinyusb/examples/device/uac2_headset$ riscv-nuclei-elf-gcc
riscv-nuclei-elf-gcc: command not found
mmormota@DESKTOP-S2K747J:/mnt/c/tinyusb/examples/device/uac2_headset$

And I double checked, that the /bin/ folder is defintely existing is the Windows path.

**Reply by cr1901** [reply]
I don't use WSL, so I'm not sure what's wrong. I'm unable to debug further I'm afraid :/. Try `/mnt/c/tinyusb/hw/mcu/gd/nuclei-sdk/gcc/bin` to the path just to see what happens...

**Reply by vorosj** [reply]
Hmmm, strange, it's not path problem. I navigate to the bin folder, dir, there is a file, OK. Then wrote the command, and command not found  :-O

mmormota@DESKTOP-S2K747J:/mnt/c/tinyusb/hw/mcu/gd/nuclei-sdk/gcc/bin$ dir
riscv-nuclei-elf-addr2line  riscv-nuclei-elf-gcc         riscv-nuclei-elf-gdb            riscv-nuclei-elf-objdump
riscv-nuclei-elf-ar         riscv-nuclei-elf-gcc-10.2.0  riscv-nuclei-elf-gdb-add-index  riscv-nuclei-elf-ranlib
riscv-nuclei-elf-as         riscv-nuclei-elf-gcc-ar      riscv-nuclei-elf-gprof          riscv-nuclei-elf-readelf
riscv-nuclei-elf-c++        riscv-nuclei-elf-gcc-nm      riscv-nuclei-elf-ld             riscv-nuclei-elf-size
riscv-nuclei-elf-c++filt    riscv-nuclei-elf-gcc-ranlib  riscv-nuclei-elf-ld.bfd         riscv-nuclei-elf-strings
riscv-nuclei-elf-cpp        riscv-nuclei-elf-gcov        riscv-nuclei-elf-lto-dump       riscv-nuclei-elf-strip
riscv-nuclei-elf-elfedit    riscv-nuclei-elf-gcov-dump   riscv-nuclei-elf-nm
riscv-nuclei-elf-g++        riscv-nuclei-elf-gcov-tool   riscv-nuclei-elf-objcopy
mmormota@DESKTOP-S2K747J:/mnt/c/tinyusb/hw/mcu/gd/nuclei-sdk/gcc/bin$  riscv-nuclei-elf-gcc
riscv-nuclei-elf-gcc: command not found
mmormota@DESKTOP-S2K747J:/mnt/c/tinyusb/hw/mcu/gd/nuclei-sdk/gcc/bin$

**Reply by vorosj** [reply]
Maybe in the WSL window it needs the **Windows** version of the toolchain? .exe files?
If I conseder that the arm toolchain in the path is a Windows version, probably this is the case. I try it...

**Reply by vorosj** [reply]
I am totally confused.
In a Win11 cmd window both the arm and the riscv gcc started, so the path has to be OK for both:
C:\>riscv-nuclei-elf-gcc
riscv-nuclei-elf-gcc: fatal error: no input files
compilation terminated.

C:\>arm-none-eabi-gcc.exe
arm-none-eabi-gcc.exe: fatal error: no input files
compilation terminated.

In the WSL window the arm started, the riscv not found:
mmormota@DESKTOP-S2K747J:/mnt/c/WINDOWS/system32$ riscv-nuclei-elf-gcc
riscv-nuclei-elf-gcc: command not found
mmormota@DESKTOP-S2K747J:/mnt/c/WINDOWS/system32$ arm-none-eabi-gcc.exe
arm-none-eabi-gcc.exe: fatal error: no input files
compilation terminated.
mmormota@DESKTOP-S2K747J:/mnt/c/WINDOWS/system32$

**Reply by vorosj** [reply]
I tried a really ugly workaround: copied all the riscv toolchain files into the arm bin folder. 
This path is working in the WSL window:
mmormota@DESKTOP-S2K747J:/mnt/c/tinyusb/examples/device/uac2_headset$ riscv-nuclei-elf-gcc.exe
riscv-nuclei-elf-gcc.exe: fatal error: no input files
compilation terminated.
Ok, that is a step forward. But then coming the usual fail:

mmormota@DESKTOP-S2K747J:/mnt/c/tinyusb/examples/device/uac2_headset$ make CROSS_COMPILE=riscv-nuclei-elf- BOARD=sipeed_longan_nano all
AS startup_gd32vf103_asm.o
make: riscv-nuclei-elf-gcc: Command not found
make: *** [../../rules.mk:143: _build/sipeed_longan_nano/obj/hw/mcu/gd/nuclei-sdk/SoC/gd32vf103/Common/Source/GCC/startup_gd32vf103_asm.o] Error 127

Linux and makefiles killing me. :-)

**Reply by vorosj** [reply]
I think the difference is: I wrote the command  riscv-nuclei-elf-gcc.**exe** and make without the .exe.

**Reply by vorosj** [reply]
Restart windows solved all the problems... :-(
The WSL window missed the new path until restart, while Win11 immediately recognised it.
I removed the ugly copies, added the LInux gcc version, and success. 

Sorry for all the lamer questions, and thank you all for help.
